Tetris2D.cpp

```diff
@@ -0,0 +1,45 @@
+#include <iostream>
+#include <cmath>
+using namespace std;
+
+const int N = 1e6+5;
+
+int sq;
+int tab[N] = {0};
+int kubelki[1005] = {0};
+
+int main() {
+    ios_base::sync_with_stdio(0);
+    cin.tie(0);
+    int d, n, ls, rs, ultram = 0;//zmienna ultram będzie trzymać maksymalną wysokość
+    cin >> d >> n;
+    sq = (int) sqrt(d + .0) + 1;
+    for(int i = 0; i<n; i++){
+        cin >> ls >> rs;
+        int l = rs, r = rs+ls-1; // żeby uniknąć problemu gdy spadają bokami na to samo miejsce numerować "pomiędzy" kreskami
+                                // od 0 do d-1
+        int blok_l = l/sq, blok_r = r/sq, m = 0;
+        //m służy do przechowania największego indeksu na który spadnie rozpatrywany klocek
+        if(blok_l==blok_r){ // w tym samym bloku
+            for(int j = l; j<=r; j++) m = max(m, tab[j]);
+            m++; // nowa wartość po dołożeniu obecnego klocka
+            for(int j = l; j<=r; j++) tab[j] = m; //zwiekszanie
+            kubelki[blok_l] = m; // tak samo z obecnym kubełkiem
+        }
+        else{
+            //to co w 1, tylko inne kubełki
+            for(int j = l, end = (blok_l+1)*sq-1; j<=end; j++) m = max(m, tab[j]); // lewy ogonek
```

Comment:
Szkoda, że nie skorzystałeś z sugestii, żeby napisać funkcję, zwracającą kubełek w którym jesteś.
Wtedy nie zamiast wzorków w for-ze mógłbyś napisać:
int j = l;
while(blok(l) == blok(j)) m = max(m, tab[j++]);